### PR TITLE
mender-sdimg: introduce MENDER_STORAGE_TOTAL_SIZE_MB, calculate rootfs size

### PR DESCRIPTION
**Pull request by bboozzoo**
Introduce MENDER_STORAGE_TOTAL_SIZE_MB that specifies the size of the
medium that will hold sdimg image. Rootfs is sized automatically by
distributing the remaining space after subtracting boot & data partition
sizes (incl. partitioning overhader) from the total size.

Generated kickstart (boot 16MB, rootfs 2 * 424, data auto sized):
```
part /boot   --source bootimg-partition --ondisk mmcblk0 --fstype=vfat --label boot --align 8192 --active --size 16
part /       --source rootfs --ondisk mmcblk0 --fstype=ext4 --label primary --align 8192 --reserved-size 424
part         --source rootfs --ondisk mmcblk0 --fstype=ext4 --label secondary --align 8192 --reserved-size 424
part /data   --source fsimage --sourceparams=file="/home/mborzecki/yocto/heads/mender-bbb/tmp/work/vexpress_qemu-poky-linux-gnueabi/core-image-minimal/1.0-r0/data.ext4" --ondisk mmcblk0 --fstype=ext4 --label data --align 8192
```

Partitions:
```
Disk /home/mborzecki/yocto/heads/mender-bbb/tmp/deploy/images/vexpress-qemu/core-image-minimal-vexpress-qemu-20161018070558.sdimg: 1008MiB
Sector size (logical/physical): 512B/512B
Partition Table: msdos
Disk Flags: 

Number  Start    End      Size     Type      File system  Flags
 1      8,00MiB  24,0MiB  16,0MiB  primary   fat16        boot, lba
 2      24,0MiB  448MiB   424MiB   primary   ext4
 3      448MiB   872MiB   424MiB   primary   ext4
 4      880MiB   1008MiB  128MiB   extended               lba
 5      880MiB   1008MiB  128MiB   logical   ext4
```

The change depends on functionality that is not yet in wic. We should wait for http://lists.openembedded.org/pipermail/openembedded-core/2016-October/127634.html to be merged into master before merging this PR.

@kacf @pasinskim @GregorioDiStefano

### CODE REVIEW

File: classes/mender-sdimg.bbclass [hunk 1]
```diff
@@ -38,6 +43,17 @@ MENDER_BOOT_PART_SIZE_MB ?= "16"
...
+# Estimate how much space may be lost due to partitioning alignment. Use a
+# simple heuristic for now - 4 partitions * alignment
+def get_overhead(d):
+    align = d.getVar('MENDER_PARTITION_ALIGNMENT_MB', True)
+    if align:
+        return 4 * int(align)
+    return 0
+
+# Overhead lost due to partitioning.
+MENDER_PARTITIONING_OVERHEADER_MB ?= "${@get_overhead(d)}"
```
Comment (bboozzoo) [inline classes/mender-sdimg.bbclass]: yup

**Comment by bboozzoo** [comment]
Well the discussion with upstream was a bit longer than I expected. Anyways, the compromise is to add `--max-size <some-mbytes>` flag to wic partitions. Means that I'll need to update the patch to reflect that.

For morty we could fall back to the pre-wic method, that is to create rootfs paritions manually and making them as large as ROOTFS_SIZE and pull each as `--fsimage`. So, the partistions will still be larger than required and larger than future update images, something I don't like personaly, because it's inconsistent. We have the fixed size advantage, so all is in good from update process' perspective.

**Comment by bboozzoo** [comment]
Refactored the code to be independent of upstream. Instead of passing a special parameter to `wic`, an image of calculated size is generated using dd. `wic` is used only for assembling particular images into a single disk image.

**Comment by kacf** [comment]
This also requires documentation here: https://github.com/mendersoftware/mender-docs/blob/master/03.Artifacts/06.Variables/docs.md

**Comment by kacf** [comment]
Running in Jenkins.

**Comment by bboozzoo** [comment]
PR to docs: https://github.com/mendersoftware/mender-docs/pull/42

**Comment by lolsborn** [comment]
This is a super useful feature.  What are the odds of getting it merged into Krogoth?

**Comment by kacf** [comment]
This is going into krogoth, for sure. See my reply on [the mailing list](https://groups.google.com/a/lists.mender.io/d/msg/mender/nAwocFiQ88M/r6fjEWrtCAAJ).

**Pull request merged**